### PR TITLE
Implement issue and pull request templates

### PR DESCRIPTION
**Pull request by andreynering**
GitHub just announced [issue and pull request templates](https://github.com/blog/2111-issue-and-pull-request-templates).

This PR implements it in Gogs and also adds templates to be used on GitHub.

### CODE REVIEW

**Comment by Unknwon** [comment]
Thanks, but I suggest split this into two PRs.

**Comment by andreynering** [comment]
@Unknwon  Ok, done.

**Comment by andreynering** [comment]
@Unknown We can open the possibility of permitting templates. inside `.gogs` folder (and maybe also `.github` folder, for compatibility). What do you think?

**Comment by Unknwon** [comment]
> possibility of permitting templates. inside .gogs folder (and maybe also .github folder, for compatibility). What do you think?

Yes, that's what I'm thinking, too

**Comment by Unknwon** [comment]
The priority would be `root -> .gogs -> .github`

**Comment by Eriner** [comment]
> The priority would be `root -> .gogs -> .github`

I really like this; will help keep folders clean. Nice to include `.github` as well for compatibility reasons. :+1:

**Comment by andreynering** [comment]
@Unknwon  Updated. It is working.

There's just a little detail, but I don't thing it is an issue: when opening a PR, the template will be taken from the user repo, not the base one. So it won't show if the user didn't update his repo. Since this is often made before starting working on something, I don't think it will be a problem.

File: routers/repo/pull.go [hunk 1]
```diff
@@ -540,6 +540,9 @@ func CompareAndPullRequest(ctx *middleware.Context) {
 	ctx.Data["PageIsComparePull"] = true
 	ctx.Data["IsDiffCompare"] = true
 	ctx.Data["RequireHighlightJS"] = true
+	setTemplateIfExists(ctx, ".github/PULL_REQUEST_TEMPLATE.md", "PullRequestTemplate")
```
Comment (Unknwon) [inline routers/repo/pull.go]: Why would you call 3 times in the handler? Shouldn't we make a function that make 3 attempts and stop once found?

Comment (andreynering) [inline routers/repo/pull.go]: You are right. I did bit of refactoring. Updated.

File: routers/repo/issue.go [hunk 2]
```diff
@@ -281,9 +283,61 @@ func RetrieveRepoMetas(ctx *middleware.Context, repo *models.Repository) []*mode
...
+	for _, filename := range possibleFiles {
+		content, found := getFileContentFromDefaultBranch(ctx, filename)
+		if found {
+			ctx.Data["IssueTemplate"] = content
+			return
+		}
+	}
+}
+
+func setPullRequestTemplateIfExists(ctx *middleware.Context) {
```
Comment (Unknwon) [inline routers/repo/issue.go]: `setIssueTemplateIfExists` and `setPullRequestTemplateIfExists` can be abstracted again.

Comment (andreynering) [inline routers/repo/issue.go]: Done.

File: routers/repo/issue.go [hunk 3]
```diff
@@ -34,11 +36,19 @@ const (
...
 	MILESTONE_EDIT base.TplName = "repo/issue/milestone_edit"
+
+	ISSUE_TEMPLATE_KEY = "IssueTemplate"
 )
 
 var (
 	ErrFileTypeForbidden = errors.New("File type is not allowed")
 	ErrTooManyFiles      = errors.New("Maximum number of files to upload exceeded")
+
+	ISSUE_POSSIBLE_TEMPLATE_FILES = []string{
```
Comment (Unknwon) [inline routers/repo/issue.go]: ALL_CAPS is not the coding style for variables...

File: routers/repo/pull.go [hunk 4]
```diff
@@ -26,6 +26,16 @@ const (
 	COMPARE_PULL base.TplName = "repo/pulls/compare"
 	PULL_COMMITS base.TplName = "repo/pulls/commits"
 	PULL_FILES   base.TplName = "repo/pulls/files"
+
+	PULL_REQUEST_TEMPLATE_KEY = "PullRequestTemplate"
+)
+
+var (
+	PULL_REQUEST_POSSIBLE_TEMPLATE_FILES = []string{
```
Comment (Unknwon) [inline routers/repo/pull.go]: Save as above. `PullReuqestTemplateCandidates` might be better?

File: routers/repo/issue.go (hunk 3 above)
Comment (andreynering) [inline routers/repo/issue.go]: Ops. Fixed.

**Comment by Unknwon** [comment]
Thanks!

**Pull request merged**